src/setup_payload/tests/TestAdditionalDataPayload.cpp

```diff
@@ -0,0 +1,283 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements a unit test suite for the additional payload generation
+ *      code functionality.
+ *
+ */
+
+#include <nlunit-test.h>
+#include <stdio.h>
+
+#include <lib/support/BytesToHex.h>
+#include <lib/support/CHIPMem.h>
+#include <lib/support/CHIPPlatformMemory.h>
+#include <setup_payload/AdditionalDataPayloadGenerator.h>
+#include <setup_payload/AdditionalDataPayloadParser.h>
+#include <setup_payload/SetupPayload.h>
+#include <system/SystemPacketBuffer.h>
+
+#include <lib/support/UnitTestRegistration.h>
+#include <lib/support/verhoeff/Verhoeff.h>
+
+using namespace chip;
+
+namespace {
+
+constexpr char kSerialNumber[]                                           = "89051AAZZ236";
+constexpr char kAdditionalDataPayloadWithoutRotatingDeviceId[]           = "1518";
+constexpr char kAdditionalDataPayloadWithRotatingDeviceId[]              = "153000120A001998AB7130E38B7E9A401CFE9F7B79AF18";
+constexpr char kAdditionalDataPayloadWithInvalidRotatingDeviceIdLength[] = "153000FF0A001998AB7130E38B7E9A401CFE9F7B79AF18";
+constexpr char kAdditionalDataPayloadWithLongRotatingDeviceId[]          = "153000130A00191998AB7130E38B7E9A401CFE9F7B79AF18";
+constexpr char kAdditionalDataPayloadWithRotatingDeviceIdAndMaxLifetimeCounter[] = "15300012FFFFFC1670A9F9666D1C4587FCBC4811549018";
+constexpr char kRotatingDeviceId[]                                               = "0A001998AB7130E38B7E9A401CFE9F7B79AF";
+constexpr uint16_t kLifetimeCounter                                              = 10;
+constexpr uint16_t kAdditionalDataPayloadLength                                  = 51;
+
+CHIP_ERROR GenerateAdditionalDataPayload(nlTestSuite * inSuite, uint16_t lifetimeCounter, const char * serialNumberBuffer,
+                                         size_t serialNumberBufferSize, BitFlags<AdditionalDataFields> additionalDataFields,
+                                         char * additionalDataPayloadOutput)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::System::PacketBufferHandle bufferHandle;
+
+    err = AdditionalDataPayloadGenerator().generateAdditionalDataPayload(
+        lifetimeCounter, serialNumberBuffer, serialNumberBufferSize, bufferHandle, additionalDataFields);
+    if (err == CHIP_NO_ERROR)
+    {
+        NL_TEST_ASSERT(inSuite, !bufferHandle.IsNull());
+    }
```

Comment:
Why is the code in the unit test repeating logic written elsewhere?

Reply:
i don't think we're repeating the code, this method is called from bluez helper which doesn't serialize the output to hex string.
we're using BytesToUppercaseHexString with a different method (i.e. generateRotatingDeviceIdAsBinary)